packages/query/src/helpers/identify-txs.ts

```diff
@@ -0,0 +1,137 @@
+        default:
+          return;
+      }
+    })
+    .filter(isDefined);
+};
+
+export const getNullifiersFromActions = (tx: Transaction): Nullifier[] => {
+  if (!tx.body?.actions) {
+    return [];
+  }
+
+  return tx.body.actions
+    .flatMap(({ action }) => {
+      switch (action.case) {
+        case 'spend':
+        case 'swapClaim':
+          return action.value.body?.nullifier;
+        default:
+          return;
+      }
+    })
+    .filter(isDefined);
+};
+
+export interface RelevantTx {
+  id: TransactionId;
+  data: Transaction;
+}
+
+type RecoveredSourceRecords = (SpendableNoteRecord | SwapRecord)[];
+
+const generateTxId = async (tx: Transaction): Promise<TransactionId> => {
+  return new TransactionId({ inner: await sha256Hash(tx.toBinary()) });
+};
+
+const searchRelevant = async (
+  tx: Transaction,
+  spentNullifiers: Set<Nullifier>,
+  commitmentRecords: Map<StateCommitment, SpendableNoteRecord | SwapRecord>,
+): Promise<
+  { relevantTx: RelevantTx; recoveredSourceRecords: RecoveredSourceRecords } | undefined
+> => {
+  let txId: TransactionId | undefined; // If set, that means this tx is relevant and should be returned to the caller
+  const recoveredSourceRecords: RecoveredSourceRecords = [];
+
+  const txNullifiers = getNullifiersFromActions(tx);
+  for (const spentNullifier of spentNullifiers) {
+    if (txNullifiers.some(txNullifier => spentNullifier.equals(txNullifier))) {
+      txId ??= await generateTxId(tx);
+    }
+  }
+
+  const txCommitments = getCommitmentsFromActions(tx);
+  for (const [stateCommitment, spendableNoteRecord] of commitmentRecords) {
+    if (txCommitments.some(txCommitment => stateCommitment.equals(txCommitment))) {
+      txId ??= await generateTxId(tx);
+
+      // Blank sources can be recovered by associating them with the transaction
+      if (BLANK_TX_SOURCE.equals(spendableNoteRecord.source)) {
+        const recovered = spendableNoteRecord.clone();
+        recovered.source = new CommitmentSource({
+          source: { case: 'transaction', value: { id: txId.inner } },
+        });
+        recoveredSourceRecords.push(recovered);
+      }
+    }
+  }
+
+  if (txId) {
+    return {
+      relevantTx: { id: txId, data: tx },
+      recoveredSourceRecords,
+    };
+  }
+
+  return undefined;
+};
+
+// identify transactions that involve a new record by comparing nullifiers and state commitments
+// also returns records with recovered sources
+export const identifyTransactions = async (
+  spentNullifiers: Set<Nullifier>,
+  commitmentRecords: Map<StateCommitment, SpendableNoteRecord | SwapRecord>,
+  blockTx: Transaction[],
+): Promise<{
+  relevantTxs: RelevantTx[];
+  recoveredSourceRecords: RecoveredSourceRecords;
+}> => {
+  const relevantTxs: RelevantTx[] = [];
+  const recoveredSourceRecords: RecoveredSourceRecords = [];
+
+  const searchPromises = blockTx.map(tx => searchRelevant(tx, spentNullifiers, commitmentRecords));
+  const results = await Promise.all(searchPromises);
```

Comment:
New feature: we now process transactions in parallel versus the previous synchronous loop

Reply:
Is there a sense of how the concurrency performs when handling large blocks with many transactions? Is the reduction in unnecessary computations and storage operations pronounced when focusing on processing only relevant transactions?

Reply:
> handling large blocks with many transactions

This should perform better as it's no longer sequentially peforming `await sha256Hash(tx.toBinary())`, but it's happening in parallel. However, the hashing cost may be trivial and this optimization may only be detectable with a very large amount of txs. 

> impact of avoiding unnecessary computations and storage operations

This code impacts this by making the storage of irrelevant transactions unnecessary. In a situation with quite large blocks with a ton of transactions, this would likely be noticeable at some point had we not fixed it.